aten/src/ATen/native/quantized/cpu/qadd.cpp

```diff
@@ -67,9 +70,83 @@ Tensor _add_scalar_out(Tensor& out, const Tensor& self, Scalar other) {
 
 template <bool ReLUFused = false>
 class QAdd final : public c10::OperatorKernel {
+#ifdef USE_PYTORCH_QNNPACK
+Tensor qnnpack_add(Tensor qa, Tensor qb, double scale, int64_t zero_point) {
+  TORCH_CHECK(qa.ndimension() > 0, "qnnpack_add(): Got empty input tensor.");
+  Tensor qa_contig = qa.contiguous();
+  Tensor qb_contig = qb.contiguous();
+
+  const auto a_zero_point = qa_contig.q_zero_point();
+  const auto b_zero_point = qb_contig.q_zero_point();
+  const auto a_scale = qa_contig.q_scale();
+  const auto b_scale = qb_contig.q_scale();
+
+  Tensor qy = at::_empty_affine_quantized(
+      qa_contig.sizes(), at::device(kCPU).dtype(kQUInt8), scale, zero_point);
+
+  if (qa_contig.size(0) == 0) {
+    return qy;
+  }
+
+  initQNNPACK();
+
+  pytorch_qnnp_operator_t qnnpack_operator{nullptr};
+
+  size_t num_elems = qa_contig.numel() / qa_contig.size(0);
+
+  const pytorch_qnnp_status createStatus = pytorch_qnnp_create_add_nc_q8(
+      num_elems /* input size */,
+      a_zero_point /* a zero_point */,
+      a_scale /* a scale */,
+      b_zero_point /* b zero_point */,
+      b_scale /* b scale */,
+      static_cast<uint8_t>(zero_point) /* sum zero_point */,
+      scale /* sum scale */,
+      std::numeric_limits<uint8_t>::min() /* output min */,
+      std::numeric_limits<uint8_t>::max() /* output max */,
+      0 /* flags */,
+      &qnnpack_operator);
+
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK Add operator");
+  TORCH_INTERNAL_ASSERT(qnnpack_operator != nullptr);
+
+  std::unique_ptr<pytorch_qnnp_operator, QnnpackOperatorDeleter>
+      qnnpack_uniq_ptr(qnnpack_operator);
+
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_add_nc_q8(
+      qnnpack_operator /* add op */,
+      qa_contig.size(0) /* batch size */,
+      (uint8_t*)qa_contig.data_ptr<c10::quint8>() /* a data */,
+      num_elems /* A stride */,
+      (uint8_t*)qb_contig.data_ptr<c10::quint8>() /* b data */,
+      num_elems /* B stride */,
+      (uint8_t*)qy.data_ptr<c10::quint8>() /* output data */,
+      num_elems /* sum stride */);
+  TORCH_INTERNAL_ASSERT(
+      setupStatus == pytorch_qnnp_status_success,
+      "failed to setup QNNPACK Add operator");
+
+  pthreadpool_t threadpool = caffe2::mobile_threadpool();
+  const pytorch_qnnp_status runStatus =
+      pytorch_qnnp_run_operator(qnnpack_operator, threadpool);
+
+  TORCH_INTERNAL_ASSERT(
```

Comment:
I think this one might not be INTERNAL ASSERT since user might have invalid input etc.?

Also looks like L131-L137 are general to all qnnpack ops? should we refactor this to separate function?

Reply:
We check for inputs separately at the start. This is an internal assert from something going wrong inside qnnpack.
I don't think refactoring line 131-137 is necessary at this point, as eventually we will change all ops API for qnnpack to make it more functional